### ISSUE TITLE
eslint exhaustive-deps: async functions should be allowed for custom effects, take 2

### ISSUE THREAD

**Issue by whatisaphone** [facebook/react#19034]
We have a custom hook, `useAsyncEffect`, which works like `useEffect` except it accepts an async function. We would like to check the deps of this function using `exhaustive-deps`, but that lint rule is going a bit beyond its name and also checking the type of function we pass.

The `exhaustive-deps` rule can't know anything about the semantics of arbitrary third-party hooks, so I think it's overstepping its bounds a bit (at least if you take the name "exhaustive deps" literally)

React version: 16.13.1
eslint version: 7.0.0
eslint-plugin-react-hooks version: 4.0.3

## Steps To Reproduce

Source:

```
import React from 'react';
import { useAsyncEffect } from './utilities/react';

function MyComponent() {
  useAsyncEffect(async () => {
    await Promise.resolve()
  }, []);
  return <div />;
}
```

`.eslintrc.json`:

```
{
  "parserOptions": {
    "ecmaVersion": 2020,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "plugins": ["react-hooks"],
  "rules": {
    "react-hooks/exhaustive-deps": [
      "error",
      {
        "additionalHooks": "^useAsyncEffect$"
      }
    ]
  }
}
```

Then run:

```
node_modules/.bin/eslint --ext jsx src/file.jsx
```

## The current behavior

```
  5:18  error  Effect callbacks are synchronous to prevent race conditions. Put the async function inside:

<snip>

✖ 1 problem (1 error, 0 warnings)
```

## The expected behavior

There should be no lint error

**Comment by MichaelBelousov** [comment]
I'd like to bump this, custom effects using dependencies have their own problems, but if you're choosing to support an `additionalHooks` option for the linter rule, it shouldn't restrict the argument specification if it can be helped. Perhaps a fork with support for @typescript-eslint (and whatever the flow alternative is) would make the most sense.

**Comment by mbkv** [comment]
:+1: 

this rule is really overstepping its boundaries. For reference, I've designed an async effect to be without race conditions

```typescript
useAsyncEffect(async rerendered => {
  await Promise.resolve(undefined);

  if (rerendered.current) {
    setState(whatever);
  }
}, [deps]);
```

it's much less verbose than the suggested way, and does exactly the same thing

```typescript
useEffect(() => {
  let cancel = false;
  async function doStuff() {
    await Promise.resolve(undefined);

    if (!cancel) {
      setState(whatever);
    }
  }

  return () => {
    cancel = true;
  }
}, [deps]);
```

**Comment by MichaelBelousov** [comment]
I made a minimal PR for this. Ideally imo, `additionalHooks` would support any arguments for a custom hook,
with a convention that the last one is always a dependency list, but I don't think I should make that decision. This PR only stops
worrying about whether a function is async for just configured `addtionalHooks`

**Comment by Dawnthorn** [comment]
I wonder if it would make more sense to move the effect callbacks async check into a separate rule like `react-hooks/async-use-effect`? In this case we want to check the dependencies, but we don't want the async check. If they were separate rules we could configure them separately.

**Comment by cgriebel** [comment]
The suggestion to split them into 2 rules I think is ideal. It also allows suppressing one with out the other when needed. I'm forced to decide between suppressing the false positive and risk dependencies being missed or having a bunch of warnings as it currently stands.

**Comment by philting** [comment]
I also agree with the proposal for separate rules. When I turn on `exhaustive-deps` I don't really expect any messages on async usage. They seem like completely different concerns to me.

**Comment by philting** [comment]
At the moment, the rule runs a check for `isEffect` which is essentially a regex match for `Effect`. Rather than assuming that effect hooks will be named `use.*Effect`, I think the rule should be generalized in case we want to check for `async` in other effect hooks. Perhaps the rule can be something like `react-hooks/no-async-effect`, with an `additionalHooks` parameter like that in `exhaustive-deps`. Default can be `use.*Effect`.

**Comment by devshorts** [comment]
Any traction on this? This is pretty annoying

**Comment by freewind** [comment]
2 years

**Comment by philting** [comment]
#18861 would likely fix this as well

**Comment by ahoisl** [comment]
Without knowing this issue, I made a PR fixing the initial problem by adding another configuration parameter, see #24958

**Comment by davemecha** [comment]
> At the moment, the rule runs a check for `isEffect` which is essentially a regex match for `Effect`. Rather than assuming that effect hooks will be named `use.*Effect`, I think the rule should be generalized in case we want to check for `async` in other effect hooks. Perhaps the rule can be something like `react-hooks/no-async-effect`, with an `additionalHooks` parameter like that in `exhaustive-deps`. Default can be `use.*Effect`.

Based on this, I found a simple workaround for the problem. Just use a name without the term `Effect` in it. This disables the async check but keeps the missing dependency check and also the suggestions for fixes.

So by renaming `useAsyncEffect` to e.g. `useAsyncEffekt` (I'm German, so this feels quite natural), the linting behavior works as expected. (assuming `useAsyncEffekt` is correctly added to `additionalHooks` in the eslint config).

| No async warning | Missing dependency warning |
| ------------- | ------------- |
| ![Image](https://github.com/user-attachments/assets/d5473d3c-b014-4eb6-bb14-d60e4000e14c) | ![Image](https://github.com/user-attachments/assets/617dde2e-4f1c-4d95-b351-6e5744bfebd9) |

🥳🥳🥳

**Comment by FStefanni** [comment]
Hi,

just to say that this issue is still valid as today (25/8/2025).
I also have my custom hook `useAsyncEffect()` which raise the same warning.
The warning is legit **only** if the hook is actually `useEffect()`, but it makes no sense on custom hooks IMHO.
But also an option to suppress it for custom hooks would do the trick.

Regards
